### PR TITLE
[minor] make recursive package dependency optional 

### LINKED ISSUE #13766: Making the `recursive` dependency an optional feature
### Is your feature request related to a problem or challenge?

The addition of the `recursive` crate as a dependency in https://github.com/apache/datafusion/pull/13310 may cause issues for some downstream projects.

For example, in the DataFusion Comet subproject, we started to see CI failures when running with `miri`.

```
error: unsupported operation: can't call foreign function `rust_psm_stack_pointer` on OS `linux`
    --> /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/psm-0.1.24/src/lib.rs:319:14
     |
319  |     unsafe { rust_psm_stack_pointer() }
     |              ^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function `rust_psm_stack_pointer` on OS `linux`
  
```

For now, we disabled the miri checks, but we would prefer to keep them and opt out of the `recursive` feature when testing in CI.

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by buraksenn**
## Which issue does this PR close?

Closes #13766

## Rationale for this change
Adding recursive package causes issues for downstream projects

## What changes are included in this PR?
Made recursive package optional in packages that use it instead of top level dependency

## Are these changes tested?
Existing tests. Especially this one that fails when recursive is off:
```
    #[test]
    fn test_large_tree() {
        let mut item = TestTreeNode::new_leaf("initial".to_string());
        for i in 0..3000 {
            item = TestTreeNode::new(vec![item], format!("parent-{}", i));
        }

        let mut visitor =
            TestVisitor::new(Box::new(visit_continue), Box::new(visit_continue));

        item.visit(&mut visitor).unwrap();
    }
```

### CODE REVIEW

**Comment by blaginin** [comment]
should we add it to the docs?

**Comment by buraksenn** [comment]
> should we add it to the docs?

Thanks. Added to README.md file

**Review by alamb — COMMENTED** [review]
Thanks @buraksenn 

I wonder if there is some way to test this (aka add a CI test that will fail with recursive protection -- perhaps the WASM compile test 🤔 ) that doesn't fail when the feature is disabled?

That way we can ensure we have the configuration settings hooked up right

File: Cargo.toml [hunk 1]
```diff
@@ -143,7 +143,6 @@ pbjson = { version = "0.7.0" }
 prost = "0.13.1"
 prost-derive = "0.13.1"
 rand = "0.8"
-recursive = "0.1.1"
```
Comment (findepi) [inline Cargo.toml:146]: revert, the version should still be defined here

File: datafusion/common/Cargo.toml [hunk 2]
```diff
@@ -61,7 +63,7 @@ object_store = { workspace = true, optional = true }
 parquet = { workspace = true, optional = true, default-features = true }
 paste = "1.0.15"
 pyo3 = { version = "0.22.0", optional = true }
-recursive = { workspace = true }
+recursive = { version = "0.1.1", optional = true }
```
Comment (findepi) [inline datafusion/common/Cargo.toml:66]:
```suggestion
recursive = { workspace = true, optional = true }
```

same in other places

Comment (findepi) [reply datafusion/common/Cargo.toml:66]: also, define default `recursive-protection` feature and make it a default feature
add `dep:recursive` to it

File: datafusion/common/Cargo.toml [hunk 3]
```diff
@@ -36,10 +36,12 @@ name = "datafusion_common"
 path = "src/lib.rs"
 
 [features]
+default = ["recursive-protection"]
 avro = ["apache-avro"]
 backtrace = []
 pyarrow = ["pyo3", "arrow/pyarrow", "parquet"]
 force_hash_collisions = []
+recursive-protection = ["recursive"]
```
Comment (findepi) [inline datafusion/common/Cargo.toml:44]:
```suggestion
recursive-protection = ["dep:recursive"]
```

Comment (findepi) [reply datafusion/common/Cargo.toml:44]: see https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies

image

File: datafusion/expr/Cargo.toml [hunk 4]
```diff
@@ -36,6 +36,8 @@ name = "datafusion_expr"
 path = "src/lib.rs"
 
 [features]
+default = ["recursive-protection"]
+recursive-protection = ["recursive"]
```
Comment (findepi) [inline datafusion/expr/Cargo.toml:40]: `dep:`

File: datafusion/expr/Cargo.toml [hunk 5]
```diff
@@ -48,7 +50,7 @@ datafusion-functions-window-common = { workspace = true }
 datafusion-physical-expr-common = { workspace = true }
 indexmap = { workspace = true }
 paste = "^1.0"
-recursive = { workspace = true }
+recursive = { version = "0.1.1", optional = true }
```
Comment (findepi) [inline datafusion/expr/Cargo.toml:53]: use workspace version

**Review by findepi — COMMENTED** [review]
thanks for working on this!

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 6]
```diff
@@ -1293,10 +1292,10 @@ mod tests {
             .build()?;
 
         let expected = vec![
-        "Filter: t2.c < UInt32(15) OR t2.c = UInt32(688) [a:UInt32, b:UInt32, c:UInt32, a:UInt32, b:UInt32, c:UInt32]",
-        "  Inner Join: t1.a + UInt32(100) = t2.a * UInt32(2) [a:UInt32, b:UInt32, c:UInt32, a:UInt32, b:UInt32, c:UInt32]",
-        "    TableScan: t1 [a:UInt32, b:UInt32, c:UInt32]",
-        "    TableScan: t2 [a:UInt32, b:UInt32, c:UInt32]",
+            "Filter: t2.c < UInt32(15) OR t2.c = UInt32(688) [a:UInt32, b:UInt32, c:UInt32, a:UInt32, b:UInt32, c:UInt32]",
+            "  Inner Join: t1.a + UInt32(100) = t2.a * UInt32(2) [a:UInt32, b:UInt32, c:UInt32, a:UInt32, b:UInt32, c:UInt32]",
```
Comment (findepi) [inline datafusion/optimizer/src/eliminate_cross_join.rs:1296]: unrelated change?

Comment (buraksenn) [reply datafusion/optimizer/src/eliminate_cross_join.rs:1296]: I think this was caused by `cargo fmt`  when I've added feature above

File: Cargo.toml (hunk 1 above)
Comment (buraksenn) [reply Cargo.toml]: added thanks

File: datafusion/common/Cargo.toml (hunk 2 above)
Comment (buraksenn) [reply datafusion/common/Cargo.toml]: Added as described thanks

**Review by findepi — APPROVED** [review]

**Comment by findepi** [comment]
> I wonder if there is some way to test this (aka add a CI test that will fail with recursive protection -- perhaps the WASM compile test 🤔 ) that doesn't fail when the feature is disabled?

@alamb would you want to see this in this PR, or in a follow-up?

**Comment by alamb** [comment]
> > I wonder if there is some way to test this (aka add a CI test that will fail with recursive protection -- perhaps the WASM compile test 🤔 ) that doesn't fail when the feature is disabled?
> 
> @alamb would you want to see this in this PR, or in a follow-up?

Yes sorry I should have been more explicit. I  think this PR should have a test (maybe even manually at first?) to both
1. ensure that this change actually does allow the relevant subcrates to be compiled without `#recursive`
2. ensure we don't break the feature in the future (for example breaking code into new crates but not bringing the `#cfg` flags

**Review by alamb — APPROVED** [review]
I spent some time playing around with this locally, I think it does the right thing

Enabled by default:
```shell
$ cargo test  -p datafusion-common --lib 2>&1  | grep test_large_tree
test tree_node::tests::test_large_tree ... ok
```

Disabled without default features:
```shell
$ cargo test --no-default-features  -p datafusion-common --lib 2>&1  | grep test_large_tree
$
```

Can enable explicitly:
```shell
$ cargo test --no-default-features  -p datafusion-common --features=recursive-protection --lib 2>&1  | grep test_large_tree
test tree_node::tests::test_large_tree ... ok
$
```

Unfortunately, I can't really think of any feasible way to add this test to CI

Sorry for the delay reviewing / approving this @buraksenn  and thank you for the reviews @findepi 

I merged up from main and plan to merge it once the tests pass

**Comment by alamb** [comment]
Thanks again @buraksenn

**Comment by alamb** [comment]
While testing this in comet, I am pretty sure this PR didn't quite fix the problem
- https://github.com/apache/datafusion-comet/pull/1154#issuecomment-2559777474

When someone includes one of these crates transitively (e.g. by using `datafusion` the feature is activated by default -- you can't turn off default features. 

Thus, when you add datafusion as the dependency, it doesn't have the 'recursive-protect' feature so there is no way to turn the feature off in subcrates. 

I think we need to turn the feature off by default in subcrates and then control enabling/disabling at the main crate level. 

I will have a PR for this shortly

**Comment by alamb** [comment]
Here is a PR to try and fix this: https://github.com/apache/datafusion/pull/13887
